waspc/src/Wasp/Generator/DbGenerator.hs

```diff
@@ -83,42 +83,67 @@ genMigrationsDir spec =
 --       - When they previously had no entities and just added their first.
-warnIfDbSchemaChangedSinceLastMigration :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO (Maybe GeneratorWarning)
-warnIfDbSchemaChangedSinceLastMigration spec projectRootDir = do
+--     In either of those scenarios, validate against DB itself to avoid redundant warnings.
+--
+--     NOTE: As one final optimization, if they do not have a schema.prisma.wasp-last-db-concurrence-checksum but the schema is
+--     in sync with the databse, we generate that file to avoid future checks.
+warnIfDbNeedsMigration :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO (Maybe GeneratorWarning)
+warnIfDbNeedsMigration spec projectRootDir = do
   dbSchemaChecksumFileExists <- doesFileExist dbSchemaChecksumFp
-
   if dbSchemaChecksumFileExists
-    then do
-      dbSchemaFileChecksum <- hexToString <$> checksumFromFilePath dbSchemaFp
-      dbChecksumFileContents <- readFile dbSchemaChecksumFp
-      return $ warnIf (dbSchemaFileChecksum /= dbChecksumFileContents) "Your Prisma schema has changed, you should run `wasp db migrate-dev`."
-    else return $ warnIf entitiesExist "Please run `wasp db migrate-dev` to ensure the local project is fully initialized."
+    then warnIfSchemaDiffersFromChecksum dbSchemaFp dbSchemaChecksumFp
+    else
+      if entitiesExist
+        then warnIfSchemaDiffersFromDb projectRootDir
+        else return Nothing
   where
     dbSchemaFp = SP.fromAbsFile $ projectRootDir </> dbSchemaFileInProjectRootDir
-    dbSchemaChecksumFp = SP.fromAbsFile $ projectRootDir </> dbSchemaChecksumOnLastMigrateFileProjectRootDir
+    dbSchemaChecksumFp = SP.fromAbsFile $ projectRootDir </> dbSchemaChecksumOnLastDbConcurrenceFileProjectRootDir
     entitiesExist = not . null $ getEntities spec
 
-    warnIf :: Bool -> String -> Maybe GeneratorWarning
-    warnIf b msg = if b then Just $ GeneratorNeedsMigrationWarning msg else Nothing
+warnIfSchemaDiffersFromChecksum :: FilePath -> FilePath -> IO (Maybe GeneratorWarning)
+warnIfSchemaDiffersFromChecksum dbSchemaFp dbSchemaChecksumFp = do
+  dbSchemaFileChecksum <- hexToString <$> checksumFromFilePath dbSchemaFp
+  dbChecksumFileContents <- readFile dbSchemaChecksumFp
+  if dbSchemaFileChecksum /= dbChecksumFileContents
+    then return . Just $ GeneratorNeedsMigrationWarning "Your Prisma schema has changed, please run `wasp db migrate-dev` when ready."
+    else return Nothing
+
+warnIfSchemaDiffersFromDb :: Path' Abs (Dir ProjectRootDir) -> IO (Maybe GeneratorWarning)
+warnIfSchemaDiffersFromDb projectRootDir = do
+  -- NOTE: If we wanted to, we could also check that the migrations dir == db,
+  -- but a schema check should handle all most likely cases.
+  schemaMatchesDb <- DbOps.doesSchemaMatchDb projectRootDir
+  case schemaMatchesDb of
+    Just True -> do
+      -- NOTE: This is an optimization, since we know schema == db.
+      -- Writing this file prevents future redundant Prisma checks.
```

Comment:
Well it makes perfect sense, I wouldn't even call it an optimization, we know they fit so we write it!